src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/documentation/LSPDocumentationProvider.java

```diff
@@ -0,0 +1,208 @@
+ *     <li>documentation for completion item</li>
+ * </ul>.
+ */
+public class LSPDocumentationProvider extends DocumentationProviderEx implements ExternalDocumentationHandler {
+
+    private static final Parser PARSER = Parser.builder().build();
+    private static final HtmlRenderer RENDERER = HtmlRenderer.builder().build();
+
+    private static final Key<Integer> TARGET_OFFSET_KEY = new Key<>(LSPDocumentationProvider.class.getName());
+
+    @Nullable
+    @Override
+    public String getQuickNavigateInfo(PsiElement element, PsiElement originalElement) {
+        return generateDoc(element, originalElement);
+    }
+
+    @Override
+    public @Nullable PsiElement getCustomDocumentationElement(@NotNull Editor editor, @NotNull PsiFile file, @Nullable PsiElement contextElement, int targetOffset) {
+        if (contextElement != null) {
+            // Store the offset where the hover has been triggered
+            contextElement.putUserData(TARGET_OFFSET_KEY, targetOffset);
+        }
+        return super.getCustomDocumentationElement(editor, file, contextElement, targetOffset);
+    }
+
+    @Nullable
+    @Override
+    public String generateDoc(PsiElement element, @Nullable PsiElement originalElement) {
+        Project project = element.getProject();
+        if (project.isDisposed()) {
+            return null;
+        }
+        Editor editor = LSPIJUtils.editorForElement(element);
+        if (editor == null) {
+            return null;
+        }
+        List<MarkupContent> result = getMarkupContents(element, originalElement);
+        if (result == null || result.isEmpty()) {
+            return null;
+        }
+        String s = result
+                .stream()
+                .map(m -> m.getValue())
+                .collect(Collectors.joining("\n\n"));
+        return styleHtml(editor, RENDERER.render(PARSER.parse(s)));
+    }
+
+    @Nullable
+    public List<MarkupContent> getMarkupContents(PsiElement element, @Nullable PsiElement originalElement) {
+        if (element instanceof LSPPsiElementForLookupItem) {
+            // Show documentation for a given completion item in the "documentation popup" (see IJ Completion setting)
+            // (LSP textDocument/completion request)
+            return ((LSPPsiElementForLookupItem) element).getDocumentation();
+        }
+        if (originalElement == null || !Objects.equals(element.getContainingFile(), originalElement.getContainingFile())) {
+            return null;
+        }
+        Editor editor = LSPIJUtils.editorForElement(element);
+        if (editor == null) {
+            return null;
+        }
+
+        // Show documentation for a hovered element (LSP textDocument/hover request).
+        VirtualFile file = originalElement.getContainingFile().getVirtualFile();
+        if (LSPVirtualFileWrapper.hasWrapper(file)) {
+            int targetOffset = getTargetOffset(originalElement);
+            originalElement.putUserData(TARGET_OFFSET_KEY, null);
```

Comment:
TARGET_OFFSET_KEY won't be nullified if we don't enter this `if` block, it needs to be done also after it

Reply:
fixed